File: Lab2/insertionsort.py
import unittest
import os

absolute_path = os.path.abspath(__file__)
file_path = os.path.join(os.path.dirname(absolute_path), "test.txt")

def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr



class TestInsertionSort(unittest.TestCase):

    def test(self):
        with open(file_path, 'r') as f:
            line = f.readline()
            unsorted_str, sorted_str = line.split('    ')
            unsorted_arr = eval(unsorted_str)
            sorted_arr = eval(sorted_str)

        self.assertEqual(insertion_sort(unsorted_arr), sorted_arr)


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: Lab2/Insertion_sort.java
import org.junit.jupiter.api.Test;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.Arrays;
import static org.junit.jupiter.api.Assertions.assertArrayEquals;

public class InsertionSortTest {

    public static int[] insertionSort(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            int key = arr[i];
            int j = i - 1;

            while (j >= 0 && key < arr[j]) {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
        return arr;
    }

    @Test
    public void testInsertionSort() throws IOException {
        // Define the file path
        File file = new File("test.txt");

        // Read the file
        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
            String line = br.readLine();

            // Parse the input from the file
            String[] parts = line.split("\\s{4}"); // Splits on 4 spaces
            int[] unsortedArray = Arrays.stream(parts[0].replace("[", "").replace("]", "").split(","))
                                         .map(String::trim)
                                         .mapToInt(Integer::parseInt)
                                         .toArray();
            int[] sortedArray = Arrays.stream(parts[1].replace("[", "").replace("]", "").split(","))
                                       .map(String::trim)
                                       .mapToInt(Integer::parseInt)
                                       .toArray();

            // Assert the result
            assertArrayEquals(sortedArray, insertionSort(unsortedArray));
        }
    }
}
